state/protocol/badger/mutator.go

```diff
@@ -658,45 +620,204 @@ func (m *FollowerState) Finalize(ctx context.Context, blockID flow.Identifier) e
 		m.metrics.BlockSealed(sealedBlock)
 	}
 
+	// apply all queued metrics
+	for _, updateMetric := range metrics {
+		updateMetric()
+	}
+
 	return nil
 }
 
-// epochStatus computes the EpochStatus for the given block
-// BEFORE applying the block payload itself
+// epochFallbackTriggeredByFinalizedBlock checks whether finalizing the input block
+// would trigger epoch emergency fallback mode. In particular, we trigger epoch
+// fallback mode when finalizing a block B when:
+// 1. B is the first finalized block with view greater than or equal to the epoch
+//    commitment deadline for the current epoch.
+// 2. The next epoch has not been committed as of B.
+//
+// This function should only be called when epoch fallback *has not already been triggered*.
+// See protocol.Params for more details on the epoch commitment deadline.
+//
+// No errors are expected during normal operation.
+func (m *FollowerState) epochFallbackTriggeredByFinalizedBlock(block *flow.Header, epochStatus *flow.EpochStatus, currentEpochSetup *flow.EpochSetup) (bool, error) {
+
+	// 1. determine whether block B is past the epoch commitment deadline
+	safetyThreshold, err := m.Params().EpochCommitSafetyThreshold()
+	if err != nil {
+		return false, fmt.Errorf("could not get epoch commit safety threshold: %w", err)
+	}
+	epochCommitmentDeadline := currentEpochSetup.FinalView - safetyThreshold
+	blockExceedsDeadline := block.View >= epochCommitmentDeadline
```

Comment:
Subtractions of view-based quantities are generally prone to underflow. I think it would be completely fine to algebraically change the computation to use addition only, which circumvents the problem entirely: 
```suggestion
	blockExceedsDeadline := block.View+safetyThreshold >= currentEpochSetup.FinalView
```

Reply:
Addressed in https://github.com/onflow/flow-go/pull/2747